include/mbedtls/psa_util.h

```diff
@@ -228,6 +229,31 @@ static inline psa_ecc_curve_t mbedtls_psa_translate_ecc_group( mbedtls_ecp_group
     }
 }
 
+/* Translations for PK layer */
+
+static inline int mbedtls_psa_err_translate_pk( psa_status_t status )
+{
+    switch( status )
+    {
+        case PSA_SUCCESS:
+            return( 0 );
+        case PSA_ERROR_NOT_SUPPORTED:
+            return( MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE );
+        case PSA_ERROR_INSUFFICIENT_MEMORY:
+            return( MBEDTLS_ERR_PK_ALLOC_FAILED );
+        case PSA_ERROR_COMMUNICATION_FAILURE:
+        case PSA_ERROR_HARDWARE_FAILURE:
+        case PSA_ERROR_TAMPERING_DETECTED:
+            return( MBEDTLS_ERR_PK_HW_ACCEL_FAILED );
+        case PSA_ERROR_INSUFFICIENT_ENTROPY:
+            return( MBEDTLS_ERR_ECP_RANDOM_FAILED );
+        case PSA_ERROR_BAD_STATE:
+            return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+        default: /* should never happen */
+            return( MBEDTLS_ERR_PK_HW_ACCEL_FAILED );
```

Comment:
The only viable alternative I can see is to have a generic error code, e.g. `UNKNOWN_ERROR` or `UNEXPECTED_ERROR`, but that has a large impact on the library. For now, since AFAIK HW failure is the only potential cause of these errors, your solution looks good.

In my own software I will terminate the program in an "impossible" situation, but it doesn't seem to be a practice in Mbed TLS.

Reply:
Yeah, so far the theory is we're a library, so we only return error code and let the application decide it it want to terminate or not. I agree it could be debated, but overall I think that's a very reasonable design and I don't think changing it is on the table. We just need better error codes. Thankfully we now have "generic" (not module-specific) error codes, so we could add a generic `UNEXPECTED_ERROR` code and/or even `ASSERTION_FAILED` that applications could choose to detect and treat as "wow, Mbed TLS is failing so badly I better abort myself".